### PR TITLE
feat: HubPool helper script

### PR DESCRIPTION
**Pull request by pxrl**
This script can be used to:
 - Scrape arbitrary HubPool events (ProposeRootBundle, DisputeRootBundle, ...)
 - Initiate a dispute, automating the use of the bond token.

Closes ACX-1502

### CODE REVIEW

**Comment by pxrl** [comment]
Note - there are a couple of todos in the code. I might address these, but I also think it could go in as-is.

**Comment by pxrl** [comment]
Here's an example of the current output from the script. It shows a two step procedure, where the first call retrieves information about the active proposal and provides instructions on how to submit the dispute, and the second call actually submits the dispute.

In this case the caller already had enough ABT and the HubPool was an approved spender, so it did not attempt transactions for those purposes. In order cases it would handle this automatically.

```
$ yarn dispute --chainId 5
yarn run v1.22.19
$ yarn ts-node ./scripts/hubpool.ts dispute --chainId 5
$ dev/across-protocol/relayer-v2/node_modules/.bin/ts-node ./scripts/hubpool.ts dispute --chainId 5
Goerli HubPool Dispute Bond:
        address                         : 0xbe37d8a126900698B1f6784a1D5C16253643D746
        symbol                          : ABT
        amount                          : 0.010000000000000001
        balance                         : 4.99

Goerli Root Bundle Proposal:
        blockNumber                     : 9646971
        transactionHash                 : 0xae44771e358a2c0ff1bf41210b259ed956d6099d3eaf0d621f5c18e036145ee3
        poolRebalanceRoot               : 0xba29b3cc125d664992890c130d4c14659745beb8e0fcfe538b1b72fc2f00038d
        relayerRefundRoot               : 0xfe828172091b64e918b1d13027be3d31416020770a8f0603cdb6df09f5e84a6f
        slowRelayRoot                   : 0xa51863c8ca5384688a3e706cbd4130d125e592bd19dcc6d0948be53cb10d60bf
        claimedBitMap                   : 0
        proposer                        : 0x9A8f92a830A5cB89a3816e3D267CB7791c16b04D
        unclaimedPoolRebalanceLeafCount : 1
        challengePeriodEndTimestamp     : 1694005597

To dispute, re-run with the following transaction hash (WARNING: THIS *WILL* SUBMIT A DISPUTE):

        --txnHash 0xae44771e358a2c0ff1bf41210b259ed956d6099d3eaf0d621f5c18e036145ee3

Done in 10.70s.
$ yarn dispute --chainId 5 --txnHash 0xae44771e358a2c0ff1bf41210b259ed956d6099d3eaf0d621f5c18e036145ee3
yarn run v1.22.19
$ yarn ts-node ./scripts/hubpool.ts dispute --chainId 5 --txnHash 0xae44771e358a2c0ff1bf41210b259ed956d6099d3eaf0d621f5c18e036145ee3
$ dev/across-protocol/relayer-v2/node_modules/.bin/ts-node ./scripts/hubpool.ts dispute --chainId 5 --txnHash 0xae44771e358a2c0ff1bf41210b259ed956d6099d3eaf0d621f5c18e036145ee3
Goerli HubPool Dispute Bond:
        address                         : 0xbe37d8a126900698B1f6784a1D5C16253643D746
        symbol                          : ABT
        amount                          : 0.010000000000000001
        balance                         : 4.99

Goerli Root Bundle Proposal:
        blockNumber                     : 9646971
        transactionHash                 : 0xae44771e358a2c0ff1bf41210b259ed956d6099d3eaf0d621f5c18e036145ee3
        poolRebalanceRoot               : 0xba29b3cc125d664992890c130d4c14659745beb8e0fcfe538b1b72fc2f00038d
        relayerRefundRoot               : 0xfe828172091b64e918b1d13027be3d31416020770a8f0603cdb6df09f5e84a6f
        slowRelayRoot                   : 0xa51863c8ca5384688a3e706cbd4130d125e592bd19dcc6d0948be53cb10d60bf
        claimedBitMap                   : 0
        proposer                        : 0x9A8f92a830A5cB89a3816e3D267CB7791c16b04D
        unclaimedPoolRebalanceLeafCount : 1
        challengePeriodEndTimestamp     : 1694005597

Disputing Goerli HubPool proposal: 0x7025f2a75627e4d1093fdaf087eccdabd9e258b0f235c1a13e8b39e947599f90.
Disputed HubPool proposal.
Done in 35.06s.
```

File: scripts/hubpool.ts [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+  const bondToken = WETH9.connect(bondTokenAddress, hubPool.provider);
+  const [bondBalance, decimals, symbol, allowance, proposals] = await Promise.all([
+    bondToken.balanceOf(signer.address),
+    bondToken.decimals(),
+    bondToken.symbol(),
+    bondToken.allowance(signer.address, hubPool.address),
+    hubPool.queryFilter(filter, fromBlock, latestBlock.number),
+  ]);
+
+  /* Resolve the existing proposal and determine whether it can still be disputed. */
```
Comment (mrice32) [inline scripts/hubpool.ts:81]: Sorry, maybe I'm missing something, but why can't we just quickly check if a dispute is possible by checking the liveness? If we're passed liveness or it's 0, no dispute possible.

File: scripts/hubpool.ts [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+  } catch (err) {
+    return usage(args.wallet) ? NODE_SUCCESS : NODE_INPUT_ERR;
+  }
+
+  let result: boolean;
+  switch (argv[0]) {
+    case "dispute":
+      result = await dispute(args, signer);
+      break;
+    case "scrape":
```
Comment (mrice32) [inline scripts/hubpool.ts:281]: What is scrape? It wasn't clear from the comments or implementation what it's intended to do?

**Review by mrice32 — APPROVED** [review]
Looks great! Just a few minor comments!

**Comment by pxrl** [comment]
Todo: This script should be as auto-config as possible, so include some public RPC endpoints that can be used out of the box if no other preferred providers have been configured.

**Comment by pxrl** [comment]
> Todo: This script should be as auto-config as possible, so include some public RPC endpoints that can be used out of the box if no other preferred providers have been configured.

Implemented: https://github.com/across-protocol/relayer-v2/pull/922/commits/97772bf609d207cb99c78602e2aab1a72c3c833f

File: scripts/hubpool.ts [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+async function getHubPoolContract(chainId: number): Promise<Contract> {
+  const contractName = "HubPool";
+  const hubPoolChainId = resolveHubChainId(chainId);
+
+  const hubPool = getDeployedContract(contractName, hubPoolChainId);
+  const provider = new ethers.providers.StaticJsonRpcProvider(getNodeUrlList(hubPoolChainId, 1)[0]);
+  return hubPool.connect(provider);
+}
```
[the remark this answers is not in the document]
Comment (pxrl) [reply scripts/hubpool.ts]: Implemented: https://github.com/across-protocol/relayer-v2/pull/922/commits/51bdc66bda6e77c2d1765fcdf2529e20ee5cd3b3

**Pull request merged**